File: src/core/api/video.ts
/**
 * Video API - Video inference operations
 */
import { invoke } from '@tauri-apps/api/core';
import type { ApiResponse } from './types';
import type { VideoInferenceConfig } from './types';

export interface VideoMetadata {
  duration: number;
  fps: number;
  frames: number;
  width: number;
  height: number;
}

export interface InferenceSession {
  inference_id: string;
  status: 'running' | 'completed' | 'failed';
}

/**
 * Load video and get metadata
 */
export async function loadVideo(videoPath: string): Promise<ApiResponse<VideoMetadata>> {
  try {
    return await invoke('video_load', { videoPath });
  } catch (error) {
    console.error('[API] loadVideo error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Start video inference (Python backend)
 */
export async function startVideoInference(
  config: VideoInferenceConfig
): Promise<ApiResponse<InferenceSession>> {
  try {
    const result = await invoke<ApiResponse<string>>('video_inference_start', { config });
    if (result.success && result.data) {
      return {
        success: true,
        data: {
          inference_id: result.data,
          status: 'running',
        },
      };
    }
    return { success: false, error: result.error };
  } catch (error) {
    console.error('[API] startVideoInference error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Start video inference using pure Rust (NEW - optimized)
 */
export async function startRustVideoInference(
  config: VideoInferenceConfig
): Promise<ApiResponse<InferenceSession>> {
  try {
    const result = await invoke<ApiResponse<string>>('rust_video_inference_start', { config });
    if (result.success && result.data) {
      return {
        success: true,
        data: {
          inference_id: result.data,
          status: 'running',
        },
      };
    }
    return { success: false, error: result.error };
  } catch (error) {
    console.error('[API] startRustVideoInference error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Stop video inference
 */
export async function stopVideoInference(sessionId?: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('video_inference_stop', { sessionId });
  } catch (error) {
    console.error('[API] stopVideoInference error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Stop Rust video inference
 */
export async function stopRustVideoInference(sessionId?: string): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('rust_video_inference_stop', { sessionId });
  } catch (error) {
    console.error('[API] stopRustVideoInference error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Capture screenshot from video
 */
export async function captureScreenshot(
  videoPath: string,
  timestampMs: number
): Promise<ApiResponse<string>> {
  try {
    return await invoke<ApiResponse<string>>('video_capture_screenshot', {
      videoPath,
      timestampMs,
    });
  } catch (error) {
    console.error('[API] captureScreenshot error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Extract frames from video
 */
export async function extractFrames(
  videoPath: string,
  intervalMs: number
): Promise<ApiResponse<string[]>> {
  try {
    return await invoke<ApiResponse<string[]>>('video_extract_frames', {
      videoPath,
      intervalMs,
    });
  } catch (error) {
    console.error('[API] extractFrames error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Get inference results
 */
export async function getVideoInferenceResults(
  inferenceId: string
): Promise<ApiResponse<unknown>> {
  try {
    return await invoke<ApiResponse<unknown>>('video_inference_results', {
      inferenceId,
    });
  } catch (error) {
    console.error('[API] getVideoInferenceResults error:', error);
    return { success: false, error: String(error) };
  }
}

File: src-tauri/src/modules/yolo/commands/video.rs
use crate::modules::yolo::models::training::{AnnotationBox, VideoInferenceConfig};
use crate::modules::yolo::services::video::VideoInfo;
use crate::modules::yolo::services::{VideoService, VideoInferenceService};
use crate::modules::yolo::services::video_inference::VideoInfo as RustVideoInfo;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tauri::{AppHandle, Emitter, State};

#[derive(Debug, Serialize, Deserialize)]
pub struct CommandResponse<T> {
    pub success: bool,
    pub data: Option<T>,
    pub error: Option<String>,
}

impl<T> CommandResponse<T> {
    pub fn ok(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
        }
    }
    pub fn err(msg: String) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(msg),
        }
    }
}

/// Load video and return metadata (uses existing VideoService)
#[tauri::command]
pub async fn video_load(
    state: State<'_, Arc<VideoService>>,
    video_path: String,
) -> Result<CommandResponse<VideoInfo>, String> {
    match state.probe_video(&video_path).await {
        Ok(info) => Ok(CommandResponse::ok(info)),
        Err(e) => Ok(CommandResponse::err(e)),
    }
}

/// Start video inference using Python (existing implementation)
#[tauri::command]
pub async fn video_inference_start(
    app: AppHandle,
    state: State<'_, Arc<VideoService>>,
    config: VideoInferenceConfig,
) -> Result<CommandResponse<String>, String> {
    let session_id = format!("vid_{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis());

    let session_id_clone = session_id.clone();
    let session_id_clone2 = session_id.clone();
    let state_arc = Arc::clone(&state);
    let session_id_return = session_id.clone();

    // Spawn inference in background
    tokio::spawn(async move {
        let app_clone = app.clone();
        let app_clone_for_callback = app.clone();
        let callback = move |frame_idx: u32, boxes: Vec<AnnotationBox>| {
            let event = serde_json::json!({
                "session_id": session_id_clone,
                "frame": frame_idx,
                "boxes": boxes,
            });
            let _ = app_clone_for_callback.emit("video-inference-frame", event);
        };

        match state_arc.run_inference(&session_id, &config, callback).await {
            Ok(results) => {
                let _ = app_clone.emit("video-inference-complete", serde_json::json!({
                    "session_id": session_id_clone2,
                    "success": true,
                    "frames": results.len(),
                }));
            }
            Err(e) => {
                let _ = app_clone.emit("video-inference-complete", serde_json::json!({
                    "session_id": session_id_clone2,
                    "success": false,
                    "error": e,
                }));
            }
        }
    });

    Ok(CommandResponse::ok(session_id_return))
}

/// Start video inference using pure Rust (NEW - optimized)
#[tauri::command]
pub async fn rust_video_inference_start(
    app: AppHandle,
    state: State<'_, Arc<VideoInferenceService>>,
    config: VideoInferenceConfig,
) -> Result<CommandResponse<String>, String> {
    let session_id = format!("rust_vid_{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis());

    let session_id_clone = session_id.clone();
    let session_id_clone2 = session_id.clone();
    let state_arc = Arc::clone(&state);
    let session_id_return = session_id.clone();

    eprintln!("[RustVideoInference] Starting session: {}", session_id);

    // Spawn inference in background
    tokio::spawn(async move {
        let app_clone = app.clone();
        let app_clone_for_callback = app.clone();
        let callback = move |frame_idx: u32, boxes: Vec<AnnotationBox>| {
            let event = serde_json::json!({
                "session_id": session_id_clone,
                "frame": frame_idx,
                "boxes": boxes,
            });
            let _ = app_clone_for_callback.emit("rust-video-inference-frame", event);
        };

        match state_arc.run_inference(&session_id, &config, callback).await {
            Ok(results) => {
                let _ = app_clone.emit("rust-video-inference-complete", serde_json::json!({
                    "session_id": session_id_clone2,
                    "success": true,
                    "frames": results.len(),
                    "results_path": format!("{}/inference_results.json", config.output_dir),
                }));
            }
            Err(e) => {
                eprintln!("[RustVideoInference] Error: {}", e);
                let _ = app_clone.emit("rust-video-inference-complete", serde_json::json!({
                    "session_id": session_id_clone2,
                    "success": false,
                    "error": e,
                }));
            }
        }
    });

    Ok(CommandResponse::ok(session_id_return))
}

/// Stop Rust video inference
#[tauri::command]
pub async fn rust_video_inference_stop(
    state: State<'_, Arc<VideoInferenceService>>,
    session_id: Option<String>,
) -> Result<CommandResponse<()>, String> {
    if let Some(sid) = session_id {
        state.stop_inference(&sid).await;
    }
    Ok(CommandResponse::ok(()))
}

/// Stop video inference (Python)
#[tauri::command]
pub async fn video_inference_stop(
    state: State<'_, Arc<VideoService>>,
    session_id: Option<String>,
) -> Result<CommandResponse<()>, String> {
    if let Some(sid) = session_id {
        state.stop_inference(&sid).await;
    }
    Ok(CommandResponse::ok(()))
}

/// Capture screenshot from video at given timestamp
#[tauri::command]
pub async fn video_capture_screenshot(
    state: State<'_, Arc<VideoService>>,
    video_path: String,
    timestamp_ms: u64,
) -> Result<CommandResponse<String>, String> {
    let output_path = format!("/tmp/screenshot_{}.jpg", timestamp_ms);
    match state.capture_screenshot(&video_path, timestamp_ms, &output_path).await {
        Ok(path) => Ok(CommandResponse::ok(path)),
        Err(e) => Ok(CommandResponse::err(e)),
    }
}

/// Extract frames from video at given interval
#[tauri::command]
pub async fn video_extract_frames(
    state: State<'_, Arc<VideoService>>,
    video_path: String,
    interval_ms: u32,
) -> Result<CommandResponse<Vec<String>>, String> {
    let output_dir = format!("/tmp/frames_{}", std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .unwrap()
        .as_millis());
    match state.extract_frames(&video_path, interval_ms, &output_dir).await {
        Ok(frames) => Ok(CommandResponse::ok(frames)),
        Err(e) => Ok(CommandResponse::err(e)),
    }
}

/// Get inference results (placeholder - results are streamed)
#[tauri::command]
pub async fn video_inference_results(
    _state: State<'_, Arc<VideoService>>,
    _inference_id: String,
) -> Result<CommandResponse<serde_json::Value>, String> {
    // Results are delivered via events, this is a placeholder for future file-based retrieval
    Ok(CommandResponse::ok(serde_json::json!({
        "message": "Results delivered via events"
    })))
}
